### PR TITLE
Irmin.Type: Ensure that decode/encode use the pre-existing buffer it …

### PR DESCRIPTION
**Pull request by samoht**
…is provided

### CODE REVIEW

File: src/irmin/type.mli [hunk 1]
```diff
@@ -133,8 +133,8 @@ val decode_json_lexemes: 'a t -> Jsonm.lexeme list -> ('a, [`Msg of string]) res
 val to_json_string: ?minify:bool -> 'a t -> 'a to_string
 val of_json_string: 'a t -> 'a of_string
 
-val encode_bin: ?buf:bytes -> 'a t -> 'a to_string
-val decode_bin: ?exact:bool -> 'a t -> 'a of_string
+val encode_bin: ?buf:(bytes * int) -> 'a t -> 'a to_string
+val decode_bin: ?exact:bool -> ?off:int -> 'a t -> 'a of_string
```
Comment (hannesm) [inline src/irmin/type.mli]: if i understand correctly, the `encode_bin` has a pretty complex semantics:
- if `~buf` is provided (a tuple of bytes and int), the `'a t` is encoded in that buffer at given offset
- if no `buf` is provided, `'a t` is encoded into a freshly allocated buffer

for the first use: how are errors handled if the given buffer is too small to encode `'a t` into it? what is the value of returning something from this function (if it is done in-place)

I'd suggest to have two separate functions here: `encode_bin` that allocates a buffer, encodes `'a t` in there, and returns the new buffer. and a separate `encode_bin_into` that requires a buffer (and offset), encodes `'a t` into it, and returns unit. (well, still need to handle the case where the given buffer is too small, and/or return an index "how many bytes were written").

Comment (samoht) [reply src/irmin/type.mli]: There are various bits to the encode/decode API that needs to be improved: I really would like a way to have a streaming codecs for both encoding and decoding which are abstracted from the actual input and output sources. We've discussed about that with @dinosaure but there is no good way to do this at the moment (or we should maybe use angstrom / faraday).

Regarding your question: if the buffer is too small, there should be an error :-)

Comment (hannesm) [reply src/irmin/type.mli]: thanks, my main point is: mixing `encode : 'a t -> bytes` and `encode_into : (bytes * int) -> 'a t -> int` into a single function (`encode_mixed : ?buf:(bytes * int) -> 'a t -> (bytes * int)`) is imho not a good idea - at the call site either there's already a buffer (and `encode_into` can be used), or there's no buffer (and `encode` can be used).

Comment (samoht) [reply src/irmin/type.mli]: Good point, I'll make the changes before merging `irmin-pack`

**Pull request merged**